src/lib/index/wb_cache.cpp

```diff
@@ -155,48 +142,206 @@ void IndexWBCache::read_buf(bnodeid_t id, BtreeNodePtr& node, node_initializer_t
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_merge_at_left_child")) {
+            child_buf->set_crash_flag();
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_merge_at_right_child")) {
+            if (!new_node_bufs.empty()) { new_node_bufs[0]->set_crash_flag(); }
+        }
+    } else if (!freed_node_bufs.empty() && (new_node_bufs.size() == freed_node_bufs.size())) {
+        // Its a rebalance node situation
+        if (iomgr_flip::instance()->test_flip("crash_flush_on_rebalance_at_meta")) {
+            if (parent_buf->is_meta_buf()) { parent_buf->set_crash_flag(); }
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_rebalance_at_root")) {
+            if (parent_buf->is_meta_buf()) { child_buf->set_crash_flag(); }
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_rebalance_at_parent")) {
+            parent_buf->set_crash_flag();
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_rebalance_at_left_child")) {
+            child_buf->set_crash_flag();
+        } else if (iomgr_flip::instance()->test_flip("crash_flush_on_rebalance_at_right_child")) {
+            if (!new_node_bufs.empty()) { new_node_bufs[0]->set_crash_flag(); }
+        }
+    }
+}
+#endif
+
+void IndexWBCache::transact_bufs(uint32_t index_ordinal, IndexBufferPtr const& parent_buf,
+                                 IndexBufferPtr const& child_buf, IndexBufferPtrList const& new_node_bufs,
+                                 IndexBufferPtrList const& freed_node_bufs, CPContext* cp_ctx) {
+    IndexCPContext* icp_ctx = r_cast< IndexCPContext* >(cp_ctx);
+    if (parent_buf) { link_buf(parent_buf, child_buf, false /* is_sibling_link */, cp_ctx); }
 
-        chain->m_next_buffer = second;
-        second->m_wait_for_leaders.increment(1);
+#ifdef _PRERELEASE
+    set_crash_flips(parent_buf, child_buf, new_node_bufs, freed_node_bufs);
+#endif
+
+    for (auto const& buf : new_node_bufs) {
+        link_buf(child_buf, buf, true /* is_sibling_link */, cp_ctx);
+    }
+
+    for (auto const& buf : freed_node_bufs) {
+        if (!buf->m_wait_for_down_buffers.testz()) {
+            // This buffer has some down bufs depending on it. It can happen for an upper level interior node, where
+            // lower level node (say leaf) has split causing it to write entries in this node, but this node is now
+            // merging with other node, causing it to free. In these rare instances, we link this node to the new
+            // node resulting in waiting for all the down bufs to be flushed before up buf can flush (this buf is
+            // not written anyways)
+            link_buf(child_buf, buf, true /* is_sibling_link */, cp_ctx);
+        }
     }
 
-    return {second_copied, third_copied};
+    icp_ctx->add_to_txn_journal(index_ordinal, child_buf->m_up_buffer,
+                                new_node_bufs.size() ? new_node_bufs[0]->m_up_buffer : nullptr, new_node_bufs,
```

Comment:
self note for me :why is `new_node_bufs[0]->m_up_buffer` used instead child_buf?